Handle empty and non-absolute installation prefixes (#2166)
nest/nest-simulator: CMakeLists.txt

project( nest CXX C )
set( NEST_USER_EMAIL "users@nest-simulator.org" )

# this include for coloring messages

> include( Utils )

Reviewer: Maybe rename the file to `ColorMessages.cmake` and remove the comment?

Please also add a new line between this `include` and the next block of code.

# check if the given CMAKE_INSTALL_PREFIX is not empty

> if( NOT IS_ABSOLUTE ${CMAKE_INSTALL_PREFIX})

Reviewer: Following the *principle of least surprise*, I think we should just convert a non-absolute path to an absolute one (see [stackoverflow/39027269](https://stackoverflow.com/questions/39027269)).

project( nest CXX C )
set( NEST_USER_EMAIL "users@nest-simulator.org" )

include( ColorMessages )

# check if the given CMAKE_INSTALL_PREFIX is not empty
if("${CMAKE_INSTALL_PREFIX}" STREQUAL "")
  set(PathLength 0)
  abortMSG("CMAKE_INSTALL_PREFIX cannot be an empty string" "CMAKE_INSTALL_PREFIX is:${CMAKE_INSTALL_PREFIX} [length:${PathLength}]")
endif()

> # handel non-aboslute path

Reviewer: ```suggestion
# handle relative installation prefixes
```

if( NOT IS_ABSOLUTE ${CMAKE_INSTALL_PREFIX})
    # convert relative path to absolute path

>     printWarning("CMAKE_INSTALL_PREFIX isn't an absolute path")

Reviewer: ```suggestion
```

    get_filename_component(absPath ${CMAKE_INSTALL_PREFIX} ABSOLUTE BASE_DIR ${CMAKE_BINARY_DIR})
    set(CMAKE_INSTALL_PREFIX ${absPath})

>     printInfo("CMAKE_INSTALL_PREFIX is now set to ${CMAKE_INSTALL_PREFIX}")  

Reviewer: ```suggestion
    printInfo("Relative CMAKE_INSTALL_PREFIX has been converted to absolute path ${CMAKE_INSTALL_PREFIX}")  
```

project( nest CXX C )
set( NEST_USER_EMAIL "users@nest-simulator.org" )

include( ColorMessages )

> # check if the given CMAKE_INSTALL_PREFIX is not empty

Reviewer 2: This comment doesn't actually match with the if condition. Would it make more sense to check for empty string outside the `if( NOT IS_ABSOLUTE ${CMAKE_INSTALL_PREFIX})` since this condition is not a pre-requisite for checking empty string?

project( nest CXX C )
set( NEST_USER_EMAIL "users@nest-simulator.org" )

# check if the given CMAKE_INSTALL_PREFIX is not empty
if("${CMAKE_INSTALL_PREFIX}" STREQUAL "")

>   set(PathLength 0)
>   message(FATAL_ERROR "CMAKE_INSTALL_PREFIX cannot be an empty string" "CMAKE_INSTALL_PREFIX is:${CMAKE_INSTALL_PREFIX} [length:${PathLength}]")

Reviewer: ```suggestion
  message(FATAL_ERROR "CMAKE_INSTALL_PREFIX cannot be set to an empty string")
```
I think the additional text is not that helpful, as the prefix is anyway empty and the length in this situation always 0.

Reviewer 2: I agree.
